flink-runtime/src/test/java/org/apache/flink/runtime/webmonitor/threadinfo/JobVertexThreadInfoTrackerTest.java

```diff
@@ -169,20 +168,32 @@ public void testCachedStatsUpdatedAfterRefreshInterval() throws Exception {
     /** Tests that cached results are removed within the cleanup interval. */
     @Test
     public void testCachedStatsCleanedAfterCleanupInterval() throws Exception {
-        final Duration cleanUpInterval2 = Duration.ofMillis(10);
+        final Duration cleanUpInterval2 = Duration.ofMillis(1);
         final long waitingTime = cleanUpInterval2.toMillis() + 10;
 
+        // use cache recording stats so we can check later that it evicted the entry
+        Cache<JobVertexThreadInfoTracker.Key, JobVertexThreadInfoStats> vertexStatsCache =
+                CacheBuilder.newBuilder()
+                        .concurrencyLevel(1)
+                        .expireAfterAccess(cleanUpInterval2.toMillis(), TimeUnit.MILLISECONDS)
+                        .recordStats()
+                        .build();
         final JobVertexThreadInfoTracker<JobVertexThreadInfoStats> tracker =
                 createThreadInfoTracker(
-                        cleanUpInterval2, STATS_REFRESH_INTERVAL, threadInfoStatsDefaultSample);
-        doInitialRequestAndVerifyResult(tracker);
+                        cleanUpInterval2,
+                        STATS_REFRESH_INTERVAL,
+                        vertexStatsCache,
+                        threadInfoStatsDefaultSample);
 
-        // wait until we are ready to cleanup
+        // no stats yet, but the request triggers async collection of stats
+        assertFalse(tracker.getVertexStats(JOB_ID, EXECUTION_JOB_VERTEX).isPresent());
+        // wait until the result is available
+        tracker.getResultAvailableFuture().get();
+        // wait for the cleanup time plus a little buffer
         Thread.sleep(waitingTime);
```

Comment:
We should try not to rely on custom timeouts, these usually result in flaky tests and make tests slower in general.

In this case, we can simply register a custom `removalListener` for guava cache, and await on `CountDownLatch`

eg.
```java
    private static class LatchRemovalListener<K, V> implements RemovalListener<K, V> {

        private final CountDownLatch latch;

        private LatchRemovalListener(CountDownLatch latch) {
            this.latch = latch;
        }

        @Override
        public void onRemoval(RemovalNotification<K, V> removalNotification) {
            latch.countDown();
        }
    }
```

Reply:
Thanks for the suggestion! I was unhappy about the various wait times and time-dependent checks, but didn't know a proper way to handle it.